apps/fft/fft_generator.cpp

```diff
@@ -0,0 +1,140 @@
+#include "Halide.h"
+
+#include "fft.h"
+
+namespace {
+
+using namespace Halide;
+
+enum class FFTNumberType { Real, Complex };
+std::map<std::string, FFTNumberType> fft_number_type_enum_map() {
+     return { { "real", FFTNumberType::Real },
+              { "complex", FFTNumberType::Complex } };
+}
+
+// Direction of FFT. Samples can be read as "time" or "spatial" depending
+// on the meaning of the input domain.
+enum class FFTDirection { SamplesToFrequency, FrequencyToSamples };
+std::map<std::string, FFTDirection> fft_direction_enum_map() {
+  return { { "samples_to_frequency", FFTDirection::SamplesToFrequency },
+           { "frequency_to_samples", FFTDirection::FrequencyToSamples } };
+}
+
+class FFTGenerator : public Halide::Generator<FFTGenerator> {
+public:
+
+    // Gain to apply to the FFT. This is folded into gains already
+    // being applied to the FFT. A gain of 1.0f indicates an
+    // unnormalized FFT. 1 / sqrt(N) gives a unitary transform 1/N
+    // gives a normalized transform for the forward direction. (Where
+    // "N" is the number of output elements of the FFT, the product of
+    // the dimensions for a complex FFT and the product of all the
+    // dimensions except the last and the last divided by 2 all minus
+    // 1 for a real FFT.)
+    GeneratorParam<float> gain{"gain", 1.0f};
+
+    // The following option specifies that a particular vector width should be
+    // used when the vector width can change the results of the FFT.
+    // Some parts of the FFT algorithm use the vector width to change the way
+    // floating point operations are ordered and grouped, which causes the results
+    // to vary with respect to the target architecture. Setting this option forces
+    // such stages to use the specified vector width (independent of the actual
+    // architecture's vector width), which eliminates the architecture specific
+    // behavior.
+    GeneratorParam<int32_t> vector_width{"vector_width", 0};
+
+    // The following option indicates that the FFT should parallelize within a
+    // single FFT. This only makes sense to use on large FFTs, and generally only
+    // if there is no outer loop around FFTs that can be parallelized.
+    GeneratorParam<bool> parallel{"parallel", false};
+
+    // Indicates forward or inverse Fourier transform --
+    // samples_to_frequency maps to a sign of -1 in the FFT and
```

Comment:
I'd remove the bits about mapping to sign, it's mostly an implementation detail since you aren't exposing the sign as a parameter like I did. I'd just say it maps to a forward or inverse FFT..